dolt/dolthub#1430: Add `--filter` option for dolt diff

codeaucafe opened the pull request:
## Summary

### TL;DR

There was no action on the original #3499 for issue #1430; the PR was closed ~3 years ago. This PR fixes the open PR comments and updates the implementation details a bit for the RowWriting of filtered rows


### Description

Adds `--filter` option to `dolt diff` for filtering by change type (`added`, `modified`, `renamed`, `dropped`, or `removed` as alias for `dropped`); please note I added renamed and dropped alias to this filter list. Completes PR #3499 by fixing the issues that blocked it from merging, plus fixes an additional bug where empty table headers were
printed when all rows were filtered out.

Users reviewing large diffs often need to focus on specific change types - deletes may need extra scrutiny while inserts are routine. With diffs spanning thousands of rows across multiple tables, grep isn't enough since updates show
 both additions and deletions.

## Implementation

**Core changes:**
- Added `diffTypeFilter` struct with validation methods in `diff.go`
- Added `--filter` argument in `arg_parser_helpers.go`
- Table filtering uses `delta.SchemaChange` and `delta.IsRename()` for modification detection
- Row filtering checks `diff.ChangeType` in `writeDiffResults()`
- Fixed control flow - uses `continue` instead of `return nil` so all tables get processed

**Empty header bug fix:**
- Implemented `lazyRowWriter` that delays `BeginTable()` until first row write
- Added `shouldUseLazyHeader()` helper to determine when lazy initialization is needed
- Only applies to filtered data-only diffs (no schema changes or renames)
- Prevents empty table headers from printing when all rows are filtered out

**Tests:**
- `go/cmd/dolt/commands/diff_filter_test.go` - comprehensive unit tests:
  - `TestShouldUseLazyHeader` - validates lazy header logic (8 test cases)
  - `TestLazyRowWriter_*` - tests lazy initialization behavior (4 test functions)
  - Mock implementations of `diffWriter` and `SqlRowDiffWriter` interfaces
- BATS integration tests for all filter types with tables, rows, and schema changes
- Edge cases: invalid values, typos, case sensitivity

## What Was Broken in PR #3499

The original PR had three blocking issues from [code review comments](https://github.com/dolthub/dolt/pull/3499) that I fixed here:

1. **No row-level filtering** - only filtered tables, not individual rows
   - Fixed by adding filtering in `writeDiffResults()`

2. **Incomplete modification detection** - didn't check for schema changes properly
   - Fixed by using `delta.SchemaChange || delta.IsRename()` as requested

3. **Loop exits early** - `return nil` prevented processing remaining tables
   - Fixed by using `continue` instead

## Additional Bug (bug from my initial filter changes; not existing prior) Fixed

**Empty table headers when filtering:**
When using `--filter` with data-only changes, if all rows were filtered out, dolt would still print the table diff header with no content:
```bash
# Before fix:
$ dolt diff HEAD~1 --filter=modified  # (table has only added rows)
diff --dolt a/t b/t
--- a/t
+++ b/t
(empty - no rows printed)

# After fix:
$ dolt diff HEAD~1 --filter=modified
(completely empty - no output at all)
```

This occurred because `BeginTable()` was called before row filtering. The `lazyRowWriter` pattern delays header output until we confirm at least one row will be written.

## Usage/Examples

```
dolt diff --filter=added      # new tables/rows
dolt diff --filter=modified   # schema changes, row updates
dolt diff --filter=renamed    # renamed tables
dolt diff --filter=dropped    # dropped tables, deleted rows
dolt diff --filter=removed    # alias for dropped
dolt diff HEAD~1 --filter=dropped -r sql
```

Fixes #1430

elianddb commented on go/cmd/dolt/cli/arg_parser_helpers.go:
```diff
@@ -320,6 +320,7 @@ func CreateDiffArgParser(isTableFunction bool) *argparser.ArgParser {
 		ap.SupportsString(FormatFlag, "r", "result output format", "How to format diff output. Valid values are tabular, sql, json. Defaults to tabular.")
 		ap.SupportsString(WhereParam, "", "column", "filters columns based on values in the diff.  See {{.EmphasisLeft}}dolt diff --help{{.EmphasisRight}} for details.")
 		ap.SupportsInt(LimitParam, "", "record_count", "limits to the first N diffs.")
+		ap.SupportsString("filter", "", "diff_type", "filters results based on the type of modification (added, modified, removed).")
```
Should follow the `NameParam` naming scheme

elianddb commented on go/cmd/dolt/commands/diff.go:
```diff
@@ -60,6 +60,11 @@ const (
 	SQLDiffOutput     diffOutput = 2
 	JsonDiffOutput    diffOutput = 3
+
+	FilterAdds     = "added"
```
Keep the same post-tense wording, ie `FilterAdded`

elianddb commented on go/cmd/dolt/commands/diff.go:
```diff
@@ -220,6 +358,11 @@ func (cmd DiffCmd) validateArgs(apr *argparser.ArgParseResults) errhand.VerboseE
 		return errhand.BuildDError("invalid output format: %s", f).Build()
 	}
 
+	filterValue, _ := apr.GetValue("filter")
```
You can use the const added in apr above here for these magic values.

elianddb commented on go/cmd/dolt/commands/diff.go:
```diff
@@ -268,6 +411,9 @@ func parseDiffDisplaySettings(apr *argparser.ArgParseResults) *diffDisplaySettin
 	displaySettings.limit, _ = apr.GetInt(cli.LimitParam)
 	displaySettings.where = apr.GetValueOrDefault(cli.WhereParam, "")
 
+	filterValue := apr.GetValueOrDefault("filter", "all")
```
this relates to magic apr vals

elianddb commented on go/cmd/dolt/commands/diff.go:
```diff
@@ -816,6 +962,37 @@ func diffUserTables(queryist cli.Queryist, sqlCtx *sql.Context, dArgs *diffArgs)
+				continue // Skip this table but continue processing others
+			}
+		}
+
```
Use the `TableDeltaSummary` object provided by `GetSummary()` member func in `table_deltas.go`. This may require rewriting your impl for the struct filter struct above, it also keeps the logic in one place. There should be a `DiffType` field you can make use of.

codeaucafe commented:
Thanks @elianddb. Sorry for the delay. I have not been able to take a look at your comments yet due to having to work very late last couple days and helping with some family stuff. I'll be sure to review your comments and ask you any questions I have when I get time to work on addressing your comments this weekend.

Thank you!

Cheers,
David

codeaucafe commented on go/cmd/dolt/commands/merge.go:
```diff
@@ -480,17 +480,17 @@ func calculateMergeStats(queryist cli.Queryist, sqlCtx *sql.Context, mergeStats
 			continue
 		}
-		if summary.DiffType == "added" {
+		if summary.DiffType == diff.DiffTypeAdded {
```
FYI, used the new diff type constants added here so we aren't used string literals here

codeaucafe replied:
FYI, used the new diff type constants added here so we aren't used string literals here

codeaucafe commented on go/cmd/dolt/commands/diff.go:
```diff
@@ -130,6 +134,133 @@ type diffStatistics struct {
+
+	// Callback to invoke before first write
+	// Set to nil after first call
+	onFirstWrite func() error
```
how is this looking @elianddb? 

Also, instead of having initialized just added nil check on onFirstWrite

codeaucafe reviewed (commented):
@elianddb I updated the code based on your comments. Let me know if I misunderstood something and implemented it wrong.

I believe the failing bats test are unrelated and failing for same reasons those other unrelated bats tests failed for

Thank you!

elianddb reviewed (approved):
@codeaucafe lgtm! I've left a singular comment on a test, but more for future reference.

elianddb reviewed (changes requested):
Gotta run tests first, I'll come back once they're done.

elianddb commented:
I'm submitting on my PR branch to cleanup the release notes and have the CI run correctly. Looks like everything passed though! Thank you for the contribution @codeaucafe!

codeaucafe commented on go/cmd/dolt/commands/diff_filter_test.go:
```diff
@@ -0,0 +1,548 @@
+	mockDW := &mockDiffWriter{}
+	realWriter := &mockRowWriter{}
+
+	beginTableCalled := false
```
Doh, sorry. Can't believe I missed that before submitting to review again
